### PR TITLE
[PRFC] API Docs for the whole backend

### PR DESCRIPTION
**Pull request by sennyeya**
## Summary
Support serving your Backstage instance's API docs in your Backstage instance's catalog. Inception 🤯 .
Screenshot 2023-04-26 at 3 38 02 PM

## Caveats
1. Plugins must register with this service to have their specs shown.
2. It only supports new plugin system plugins.
3. The work is only to create the actual OpenAPI spec through a provider. It was decided that exposing a merged spec as an endpoint was less important than showing it in the UI.

## Testing
Since the Uffizi instance isn't set up to serve the `backend-next` package, to see the full benefits of this package, you'll have to test locally. `cd packages/backend-next && yarn start` or `yarn --cwd packages/backend-next start` and then issue a GET request to `${your_base_url}/api/api-registry.openapi/openapi.json` (still a WIP on the URL). You should see all registered OpenAPI specs under their correct prefixes. Currently, this is just the catalog API.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: app-config.yaml
Comment (sennyeya) [inline app-config.yaml:50]: Catalog imports don't support reading an OpenAPI spec without this.

File: plugins/catalog-backend/src/service/CatalogPlugin.ts
Comment (sennyeya) [inline plugins/catalog-backend/src/service/CatalogPlugin.ts:113]: Before this merges, extension points have to be added to `CatalogPlugin` for support of this resolution.

File: plugins/search-backend/src/alpha.ts
Comment (camilaibs) [inline plugins/search-backend/src/alpha.ts:114]: Just a question from an adopter perspective, would this be something we want to give adopters the ability to disable or optionally unregister for this or any other plugin via the `app-config.yaml` file?

**Review by camilaibs — APPROVED** [review]
Looking forward to use Backstage to document our plugin APIs 🎉  , thanks for contributing!
The changes to `search-backend` looks great to me, I've added only a curiosity question below. I'll leave the registry plugin review to @backstage/maintainers , I think they'll have great inputs to share!

File: plugins/search-backend/src/alpha.ts
Comment (camilaibs) [reply plugins/search-backend/src/alpha.ts:114]: > My thinking was that adopters would want to see all installed plugins' APIs

A-ha, got it! That makes sense to me; thanks for clarifying 🙌🏻 !

**Comment by benjdlambert** [comment]
Sorry this is on my list to have a look at!

**Comment by benjdlambert** [comment]
@sennyeya looks like there's some conflicts that have crept in here :(

**Comment by sennyeya** [comment]
@benjdlambert Sorry for the delay, this should be good to review now.

**Comment by benjdlambert** [comment]
Hey @sennyeya sorry I haven't forgotten about this, just waiting for the other stuff to settle in the dependant PR's 👍

**Comment by benjdlambert** [comment]
@sennyeya do you want to update this PR with the changes now that #18102 is merged?

**Comment by sennyeya** [comment]
@benjdlambert This should be good to review!

**Comment by benjdlambert** [comment]
Hey @sennyeya, thanks for the patience here!

I'm a little unsure about having the `api-registry` a requirement to run plugins without it being a `coreService`, it would be much nicer the flip the control a little bit and have reflection to work out what plugins are installed and register them that way.

I want to chat with the other @backstage/maintainers when they come back from vacations, so gonna add the label if that's OK. Sorry about the delay again!

**Comment by sennyeya** [comment]
No worries, I was thinking similarly and had opened #17538 as an RFC to track the reflection functionality. Freben brought up the use of a more declarative plugin instead of reflection and I ran with that. Definitely open to other options.

One of the big things that I imagine is going to pop up is multiple backends and how to serve a correct spec. As is, this and a naive reflection as I had previously wouldn't handle that.

**Comment by sennyeya** [comment]
Yeah, my thinking initially was to have each plugin host their API spec on an endpoint (`my-plugin/openapi.json`) and then use the discovery API to correctly grab the plugin locations and merge the specs like that. It solves the issues of distributed discovery, but adds a weird API surface to the plugin APIs. I don't know if using the existing plugin utils (like `extensionPoint`) would work.

**Comment by Rugvip** [comment]
👋 just wanted to let you know that we're ramping up as we're mostly back from vacation and beginning to have a look at this. This one is gonna need a bit more discussion so will take a bit more time. If there's any more reasoning around this that you want to add that could be helpful.

Personally I feel there's a bit of an overlap across this an a more generic service discovery layer, which may be the way to go instead. On top of that we'd then have some OpenAPI introspection system similar to your latest comment.

**Comment by sennyeya** [comment]
@Rugvip I've also been thinking the API client generation would leverage this work for a runtime client definition. 

I definitely agree this fits into the service discovery/reflection space (see #17538). It might be a good first use case, but also might be overkill versus something more declarative.

Could we setup a call when everybody is ramped up?

**Comment by Rugvip** [comment]
@sennyeya yep can do! let's figure it out on Discord

**Comment by pjungermann** [comment]
there seem to be some changesets missing still

**Comment by sennyeya** [comment]
@Rugvip @pjungermann This should be ready for a re-review. It looks like the failing E2E tests are happening across the project and related to upstream (`swagger-ui-react` changes). 

I updated the `@backstage/backend-openapi-utils` to add the spec to a `${pluginEndpoint}/openapi.json` using `req.originalUrl` to merge the values. That's not a "perfect" solution as it uses the request path to infer what to merge onto the spec and return. Ideally, we could use the `DiscoveryApi` to get that endpoint (and servers), but I wasn't able to get to  simple solution using the current generated content in the `*.generated.ts` files.

File: plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts
Comment (sennyeya) [inline plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts:68]: How much configurability do we want here?

File: plugins/catalog-backend-module-openapi-spec/src/index.ts
Comment (tudi2d) [inline plugins/catalog-backend-module-openapi-spec/src/index.ts:49]: might want to delete this:)

File: plugins/catalog-backend-module-openapi-spec/package.json
Comment (tudi2d) [inline plugins/catalog-backend-module-openapi-spec/package.json:3]: This should probably be at least `0.1.0` to allow future changes to align with semantic versioning

File: plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts
Comment (tudi2d) [reply plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts:68]: Think this might be fine for the start - can be iterated depending on feedback.

**Review by tudi2d — CHANGES_REQUESTED** [review]
Good work - would be great to get this in :) Don't have the full context, but just noticed some small things when looking over this.

File: plugins/catalog-backend-module-openapi-spec/src/index.ts
Comment (sennyeya) [reply plugins/catalog-backend-module-openapi-spec/src/index.ts]: whoops 😅 , updated

File: plugins/catalog-backend-module-openapi-spec/catalog-info.yaml
Comment (tudi2d) [inline plugins/catalog-backend-module-openapi-spec/catalog-info.yaml:9]: should this rather be the OpenAPI Maintainers?

```suggestion
  owner: openapi-tooling-maintainers
```

**Review by tudi2d — APPROVED** [review]
thanks for adjusting - looks good to me, small nit on ownership in `catalog-info.yaml?

File: plugins/catalog-backend-module-openapi-spec/catalog-info.yaml
Comment (sennyeya) [reply plugins/catalog-backend-module-openapi-spec/catalog-info.yaml:9]: Hmm, maybe? I think this package was created before `openapi-tooling-maintainers` was a thing. @Rugvip thoughts on updating this and the codeowners?

File: plugins/catalog-backend-module-openapi-spec/README.md
Comment (alde) [inline plugins/catalog-backend-module-openapi-spec/README.md:10]: this line seems cut off

File: plugins/catalog-backend-module-openapi-spec/README.md
Comment (alde) [inline plugins/catalog-backend-module-openapi-spec/README.md:27]: does that mean that there'd have to be a `@backstage/plugin-todo-backend-module-openapi-spec` and a `@backstage/plugin-search-backend-module-openapi-spec` too?

as in, will any backend plugin need also an openapi spec module too, and it's just that catalog was a  place to start?

Comment (sennyeya) [reply plugins/catalog-backend-module-openapi-spec/README.md:27]: The naming here is a bit unfortunate as we already have a `@backstage/plugin-catalog-backend-module-openapi` that pulls in OpenAPI specs from `$ref`s. This is a module to add all of the internal registered specs into the catalog. You wouldn't need a `@backstage/plugin-todo-backend-module-openapi-spec` as long as you list `todo` or `search` in the config under `openapi.plugins`.

File: plugins/catalog-backend-module-openapi-spec/README.md
Comment (sennyeya) [reply plugins/catalog-backend-module-openapi-spec/README.md:10]: 😅 , will update

File: .changeset/gentle-pears-camp.md
Comment (sennyeya) [reply .changeset/gentle-pears-camp.md:2]: As this is the initial release, does it need a changeset? Wasn't sure...

Comment (Edje-C) [reply .changeset/gentle-pears-camp.md:2]: Yeah, I interpret this as initial releases requiring changesets
https://github.com/backstage/backstage/blob/master/REVIEWING.md#when-is-a-changeset-needed

Comment (sennyeya) [reply .changeset/gentle-pears-camp.md]: easy enough to fix! updated

**Review by Edje-C — APPROVED** [review]

**Comment by Edje-C** [comment]
Need a maintainer @Rugvip

File: packages/backend-openapi-utils/config.d.ts
Comment (Rugvip) [inline packages/backend-openapi-utils/config.d.ts:21]: This kinda makes it sound like we've got plugins for OpenAPI imo. I think it would be best if the provider follows the same config pattern as other provider, e.g. https://github.com/backstage/backstage/blob/8960902a4ea002d0a44c376f071048b0ef90053f/plugins/catalog-backend-module-github/src/providers/GithubEntityProviderConfig.ts#L50 https://github.com/backstage/backstage/blob/fa7004d9722c7f8290acaa5842af769512b7cf3e/plugins/catalog-backend-module-github/config.d.ts#L57

The equivalent of the example config in this PR would be

```yaml
catalog:
  providers:
    openapi:
      plugins:
        - catalog
        - search
        - todo
```

Jumping ahead a bit this could also allow this to be a more generic OpenAPI provider, or if that's a bad idea we could name it something more specific plugin `catalog.providers.pluginOpenapiSpec`

File: packages/backend-openapi-utils/api-report.md
Comment (Rugvip) [inline packages/backend-openapi-utils/api-report.md:252]: What do you think about only making this one public API? Skipping `OPENAPI_SPEC_ROUTE`

File: plugins/catalog-backend-module-openapi-spec/package.json
Comment (Rugvip) [inline plugins/catalog-backend-module-openapi-spec/package.json:7]: Intentional? Fine by me anyway if we consider this to be very experimental, especially given the naming discussion

File: plugins/catalog-backend-module-openapi-spec/package.json
Comment (Rugvip) [inline plugins/catalog-backend-module-openapi-spec/package.json:2]: Just to think about this name a bit, what do you think makes sense long term for this package? Would we ever add the ability to scan remote OpenAPI endpoints outside of the Backstage ecosystem? It's not particularly needed given the other `openapi` module and it likely makes more sense to add that kind of discovery provider there. With that in mind I'm wondering if something like `backstage-openapi` or `plugins-openapi` might make the naming more accurate?

File: plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts
Comment (Rugvip) [inline plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts:217]: `ANNOTATION_LOCATION` and `ANNOTATION_ORIGIN_LOCATION` for these.

I take it this entire structure is fairly experimental and likely to evolve? Otherwise I'm thinking it would make sense to make it possible to configure the name and perhaps some other fields in the entity.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Alright, another round and hopefully almost there 😁 

Sorry about the silence on this one

File: plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts
Comment (sennyeya) [reply plugins/catalog-backend-module-openapi-spec/src/InternalOpenApiDocumentationProvider.ts:217]: Yeah, for now I'd like to keep the extensability small as this is all _very_ experimental

File: plugins/catalog-backend-module-openapi-spec/package.json
Comment (sennyeya) [reply plugins/catalog-backend-module-openapi-spec/package.json:2]: I like `backstage-openapi`, my naming creativity was near 0 on this one. Will update.

File: plugins/catalog-backend-module-openapi-spec/package.json
Comment (sennyeya) [reply plugins/catalog-backend-module-openapi-spec/package.json:7]: Unintentional, as this was the default when I created, will update

**Comment by sennyeya** [comment]
@Rugvip I updated per your comments. Assuming this builds, it should be ready for another round.

File: plugins/catalog-backend-module-backstage-openapi/config.d.ts
Comment (Rugvip) [inline plugins/catalog-backend-module-backstage-openapi/config.d.ts:23]: Shipped a tweak to move the config schema over here instead, fix'd it, and renamed `openapi` -> `backstageOpenapi` to have it align with the package name and avoid confusion with a potential future `openapi` entity provider from the other module.

Comment (sennyeya) [reply plugins/catalog-backend-module-backstage-openapi/config.d.ts:23]: thanks!

**Pull request merged**
